src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -40,9 +40,16 @@ internal sealed class FileChangeWatcher
         private static readonly Func<Task<IVsAsyncFileChangeEx>, object, Task<IVsAsyncFileChangeEx>> _executeActionDelegate =
             async (precedingTask, state) =>
             {
+                // Store a local copy of the previous result before the first await, and then explicitly set the
+                // parameter to null. It's fine for the closure to GC root the IVsAsyncFileChangeEx service, but we
+                // don't want it to GC root the chain of antecedent tasks which led to this point.
```

Comment:
```suggestion
                // parameter to null. It's fine for the closure to GC root the IVsAsyncFileChangeEx service, but we
                // don't want it to GC root the chain of antecedent tasks which led to this point.
```

I'm not understanding this bit -- once the preceding task is complete it should have dropped any other state it had at that point. Put another way, we do this pattern in other places, do those also need fixing?

Reply:
> I'm not understanding this bit -- once the preceding task is complete it should have dropped any other state it had at that point.

It wasn't the `Task` that was holding references, but rather the async state machine.

Reply:
So the state machine shouldn't be held once the task is completed. And looking at the originally provided dump:

![image](https://user-images.githubusercontent.com/201340/97340697-58ff4f80-1841-11eb-8b30-489ac093abce.png)

The tasks aren't complete.

Reply:
What this looks like to me is for some reason the oldest task isn't actually making progress for some reason, and each task that is 'older' in the queue roots the newer ones, since indeed, once that task completes it needs a way to kick the next task in the queue to start working.

Reply:
It sounds like you're saying this isn't a singly-linked list of *old* tasks but rather a list of future tasks?

Reply:
And yes, if you use the "inclusive bytes" as a way to measure "how far am to the most recent task", as you can see that the further you get to the most recent task, the longer your root path becomes. In this case the root is within the FileChange service itself, since that one is is the one that's actually "running".

Reply:
@AArnott Exactly.

Reply:
@AArnott I'm not sure which tool you were using to generate that stack, is that something that tool would potentially process and display that way?

Reply:
And note this makes sense your description: if this was during shutdown, we would have queued the work to unwatch the solutions and projects that were closed by the project system earlier.

Reply:
The stack is displayed by the VS debugger's Parallel Stacks window (Tasks view). It's meant to imitate the causal stack for async state machines by looking at their continuations. So if you're correct, that picture may be showing the _oldest_ Task on "top" because when it completes, it unblocks the next Task the way an async method would resume its caller when done.

Reply:
Ah, didn't realize it had that mode these days. Handy! But yes my reading of this the chain of tasks are waiting for the one that is visually on the top to complete first.